server/src/main/java/org/elasticsearch/index/analysis/PreBuiltAnalyzerProviderFactory.java

```diff
@@ -22,41 +22,85 @@
+    private final PreBuiltAnalyzerProvider current;
 
-    public PreBuiltAnalyzerProviderFactory(String name, AnalyzerScope scope, Analyzer analyzer) {
-        analyzerProvider = new PreBuiltAnalyzerProvider(name, scope, analyzer);
+    PreBuiltAnalyzerProviderFactory(String name, PreBuiltAnalyzers preBuiltAnalyzer) {
+        super(name, new PreBuiltAnalyzersDelegateCache(name, preBuiltAnalyzer));
+        this.create = preBuiltAnalyzer::getAnalyzer;
+        current = new PreBuiltAnalyzerProvider(name, AnalyzerScope.INDICES, preBuiltAnalyzer.getAnalyzer(Version.CURRENT));
     }
 
-    public AnalyzerProvider<?> create(String name, Settings settings) {
-        Version indexVersion = Version.indexCreated(settings);
-        if (!Version.CURRENT.equals(indexVersion)) {
-            PreBuiltAnalyzers preBuiltAnalyzers = PreBuiltAnalyzers.getOrDefault(name, null);
-            if (preBuiltAnalyzers != null) {
-                Analyzer analyzer = preBuiltAnalyzers.getAnalyzer(indexVersion);
-                return new PreBuiltAnalyzerProvider(name, AnalyzerScope.INDICES, analyzer);
-            }
+    public PreBuiltAnalyzerProviderFactory(String name, Function<Version, Analyzer> create) {
+        this(name, PreBuiltCacheFactory.CachingStrategy.LUCENE, create);
+    }
+
+    public PreBuiltAnalyzerProviderFactory(String name, PreBuiltCacheFactory.CachingStrategy cache, Function<Version, Analyzer> create) {
+        super(name, cache);
+        this.create = create;
+        this.current = new PreBuiltAnalyzerProvider(name, AnalyzerScope.INDICES, create.apply(Version.CURRENT));
+    }
+
+    @Override
+    public AnalyzerProvider<?> get(IndexSettings indexSettings,
+                                   Environment environment,
+                                   String name,
+                                   Settings settings) throws IOException {
+        Version versionCreated = Version.indexCreated(settings);
+        if (Version.CURRENT.equals(versionCreated) == false) {
+            return super.get(indexSettings, environment, name, settings);
+        } else {
+            return current;
         }
+    }
 
-        return analyzerProvider;
+    @Override
+    protected AnalyzerProvider<?> create(Version version) {
+        assert Version.CURRENT.equals(version) == false;
+        return new PreBuiltAnalyzerProvider(getName(), AnalyzerScope.INDICES, create.apply(version));
     }
 
     @Override
-    public AnalyzerProvider<?> get(IndexSettings indexSettings, Environment environment, String name, Settings settings)
-            throws IOException {
-        return create(name, settings);
+    public void close() throws IOException {
+        IOUtils.close((Closeable) cache.values().stream().map(AnalyzerProvider::get));
     }
 
-    public Analyzer analyzer() {
-        return analyzerProvider.get();
+    private static class PreBuiltAnalyzersDelegateCache implements PreBuiltCacheFactory.PreBuiltCache<AnalyzerProvider<?>> {
```

Comment:
I think this is worth some javadoc and comments to explain why we do it. It makes sense but things like an empty `put` method will confuse future me when I read this class in six months.

Reply:
It'd be super nice to explain that we only need this because of `PreBuiltAnalyzers`.